### PR TITLE
circe refined pattern validation poc

### PR DESCRIPTION
**Pull request by stanislav-chetvertkov**
Adding support for `pattern: ...` regex refinement types

### CODE REVIEW

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala [hunk 1]
```diff
@@ -447,11 +449,27 @@ class CirceProtocolGenerator private (circeVersion: CirceModelGenerator, applyVa
...
       encoder: Option[scala.meta.Defn.Val],
-      decoder: Option[scala.meta.Defn.Val]
+      decoder: Option[scala.meta.Defn.Val],
+      protocolParameters: List[ProtocolParameter[ScalaLanguage]]
   ) = {
     val extraImports: List[Import] = deps.map { term =>
       q"import ${term}._"
     }
+
+    val hacks: List[Stat] = protocolParameters.map(_.propertyValidation).collect { case Some(v) => v }.distinct.map { pattern =>
```
Comment (stanislav-chetvertkov) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala:459]: need Defn.Val here

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala [hunk 2]
```diff
@@ -177,31 +177,32 @@ class CirceProtocolGenerator private (circeVersion: CirceModelGenerator, applyVa
...
           case core.DeferredMap(tpeName, customTpe) =>
             val concreteType = lookupTypeName(tpeName, concreteTypes)(identity)
             val innerType    = concreteType.getOrElse(Type.Name(tpeName))
             val tpe          = t"${customTpe.getOrElse(t"_root_.scala.Predef.Map")}[_root_.scala.Predef.String, $innerType]"
             for {
-              validatedType <- applyValidations(tpe, property)
+              validatedType <- applyValidations(tpe, property, clsName)
             } yield (validatedType, Option.empty, ReifiedRawType.ofMap(fallbackRawType))
         }
+        pattern <- Target.pure(property.downField("pattern", _.getPattern).unwrapTracker)
```
Comment (blast-hardcheese) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala:205]: Maintain `Tracker`
```suggestion
        pattern = property.downField("pattern", _.getPattern)
```

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala [hunk 3]
```diff
@@ -447,16 +449,25 @@ class CirceProtocolGenerator private (circeVersion: CirceModelGenerator, applyVa
...
       encoder: Option[scala.meta.Defn.Val],
-      decoder: Option[scala.meta.Defn.Val]
+      decoder: Option[scala.meta.Defn.Val],
+      protocolParameters: List[ProtocolParameter[ScalaLanguage]]
   ) = {
     val extraImports: List[Import] = deps.map { term =>
       q"import ${term}._"
     }
+
+    val helperRegexTypes: List[Defn.Val] = protocolParameters.flatMap(_.propertyValidation).distinct.map { pattern: String =>
```
Comment (blast-hardcheese) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala:459]: Thread through Tracker, so you can use `.history` to give good error messages if `parse[Term].get` fails
```suggestion
    val helperRegexTypes: List[Defn.Val] = protocolParameters.flatTraverse(_.propertyValidation.indexedSequence.toList).map(_.distinct).map { pattern: Tracker[String] =>
```

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala [hunk 4]
```diff
@@ -23,9 +23,15 @@ object CirceRefinedProtocolGenerator {
   def apply(circeRefinedVersion: CirceRefinedModelGenerator): ProtocolTerms[ScalaLanguage, Target] =
     fromGenerator(CirceProtocolGenerator.withValidations(circeRefinedVersion.toCirce, applyValidations _))
 
-  def applyValidations(tpe: Type, prop: Tracker[Schema[_]]): Target[Type] = {
+  def applyValidations(tpe: Type, prop: Tracker[Schema[_]], className: String): Target[Type] = {
```
Comment (blast-hardcheese) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala:26]:
```suggestion
  def applyValidations(className: String, tpe: Type, prop: Tracker[Schema[_]]): Target[Type] = {
```

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala [hunk 5]
```diff
@@ -23,9 +23,15 @@ object CirceRefinedProtocolGenerator {
...
     import scala.meta._
     tpe match {
+      case t"String" =>
+        val pattern = prop.downField("pattern", _.getPattern).unwrapTracker
+        Target.pure(pattern.fold(tpe){ pat =>
+          val refined = s"""_root_.eu.timepit.refined.string.MatchesRegex[$className.`"$pat"`.T]""".parse[Type].get
```
Comment (blast-hardcheese) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala:32]: I'm surprised this unquotes

Comment (blast-hardcheese) [reply modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala:32]: Also, when you get to the point of `Target.raiseException`, please consider `prop.downField("pattern", _.getPattern).history` in your error message

**Review by blast-hardcheese — COMMENTED** [review]
looks good so far, I'm strongly considering just adding contraband-style POJO classes that encapsulate all arguments to all framework functions, to avoid bincompat issues when adding new parameters like in this PR.

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala [hunk 6]
```diff
@@ -28,12 +29,19 @@ class CirceProtocolGeneratorLoader extends ProtocolGeneratorLoader {
 
 object CirceProtocolGenerator {
   def apply(circeVersion: CirceModelGenerator): ProtocolTerms[ScalaLanguage, Target] =
-    new CirceProtocolGenerator(circeVersion, (tpe, _) => Target.pure(tpe))
-  def withValidations(circeVersion: CirceModelGenerator, applyValidations: (Type, Tracker[Schema[_]]) => Target[Type]): ProtocolTerms[ScalaLanguage, Target] =
+    new CirceProtocolGenerator(circeVersion, WithValidations.ignore)
+  def withValidations(circeVersion: CirceModelGenerator, applyValidations: WithValidations): ProtocolTerms[ScalaLanguage, Target] =
     new CirceProtocolGenerator(circeVersion, applyValidations)
+
+  trait WithValidations {
```
Comment (stanislav-chetvertkov) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala:36]: I think a SAM interface is better from readability perspective, at least the arguments have names and one can use name = value assignment in an arbitrary order if needed making it easier to refactor

Comment (blast-hardcheese) [reply modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceProtocolGenerator.scala:36]: This is a great idea.

File: modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala [hunk 7]
```diff
@@ -27,11 +28,13 @@ object CirceRefinedProtocolGenerator {
...
-        Target.pure(pattern.fold(tpe) { pat =>
-          val refined = s"""_root_.eu.timepit.refined.string.MatchesRegex[$className.`"$pat"`.T]""".parse[Type].get
-          t"""String Refined $refined"""
-        })
+        prop.downField("pattern", _.getPattern).fold(Target.pure(tpe)) { patternTracker =>
+          Try {
+            val pat = patternTracker.unwrapTracker
+            val refined = s"""_root_.eu.timepit.refined.string.MatchesRegex[$className.`"$pat"`.T]""".parse[Type].get
+            t"""String Refined $refined"""
+          }.fold(th => Target.raiseUserError(s"$th: ${patternTracker.showHistory}"), Target.pure)
```
Comment (stanislav-chetvertkov) [inline modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala:36]: should it be Target.raiseException?

Comment (blast-hardcheese) [reply modules/scala-support/src/main/scala/dev/guardrail/generators/scala/circe/CirceRefinedProtocolGenerator.scala]: Fortunately this went away when `.parse` was removed -- Perhaps there should have been a better combinator, but the point you raise is a good one -- when to use it and how to format the error message

File: modules/sample/src/main/resources/validation/example.yaml [hunk 8]
```diff
@@ -32,6 +32,9 @@ definitions:
         format: int32
         minimum: 0
         maximum: 100
+      pattern_validation:
+        type: string
+        pattern: "[0-9]+"
```
Comment (blast-hardcheese) [inline modules/sample/src/main/resources/validation/example.yaml:37]: One lingering issue. According to [the docs](https://swagger.io/docs/specification/data-models/data-types/#pattern)...

> Without `^…$`, pattern works as a partial match, that is, matches any string that contains the specified regular expression.

It looks as though this violates the assumptions baked into refined:

```scala
scala> val pat = _root_.shapeless.Witness("[0-9]+")
pat: shapeless.Witness.Aux[String("[0-9]+")] = shapeless.Witness$$anon$1@7d875a3e

scala> val x: String Refined MatchesRegex[pat.T] = "foo1234bar"
<console>:24: error: Predicate failed: "foo1234bar".matches("[0-9]+").
       val x: String Refined MatchesRegex[pat.T] = "foo1234bar"
                                                   ^
```

Before we merge, could you look into this please?

Comment (stanislav-chetvertkov) [reply modules/sample/src/main/resources/validation/example.yaml:37]: that is a good catch, I'll prepend/append `.*` if there is no `^` or `$` accordingly

**Pull request merged**